File: chat_engine.py
"""
Chat orchestrator — takes natural language, generates DAX, executes with RLS,
and returns a natural-language answer.
"""

from __future__ import annotations

import json
import logging
import re
from typing import Any

from azure.identity import DefaultAzureCredential, get_bearer_token_provider
from openai import AsyncAzureOpenAI

from config import settings, get_rls_config
from powerbi_service import execute_dax, get_dataset_schema, get_user_filter_values, get_dax_token

logger = logging.getLogger(__name__)

_client: AsyncAzureOpenAI | None = None


def _get_openai_client() -> AsyncAzureOpenAI:
    global _client
    if _client is None:
        credential = DefaultAzureCredential()
        token_provider = get_bearer_token_provider(
            credential, "https://cognitiveservices.azure.com/.default"
        )
        _client = AsyncAzureOpenAI(
            azure_endpoint=settings.azure_openai_endpoint,
            azure_ad_token_provider=token_provider,
            api_version=settings.azure_openai_api_version,
        )
    return _client


# ---------------------------------------------------------------------------
# System prompt — built dynamically from schema + RLS config
# ---------------------------------------------------------------------------

SYSTEM_PROMPT_BASE = """\
You are a helpful data analyst chatbot. You answer questions about business data
stored in a Power BI semantic model (dataset).

## Dataset Schema
{schema_json}

{rls_context}

## Rules
1. When the user asks a data question, generate a **single valid DAX query**
   wrapped in a ```dax code fence.
2. Use EVALUATE to return a table result.  Do NOT use DEFINE unless needed for
   variables.
3. Only reference tables, columns, and measures from the schema above.
4. {rls_filter_rule}
5. If the user asks something unrelated to the data, respond conversationally
   WITHOUT generating DAX.
6. Keep DAX concise and correct.  Prefer SUMMARIZECOLUMNS for aggregations.
7. After receiving the query results, provide a clear, concise natural-language
   answer.  If the results are tabular, format them as a Markdown table.
8. If a query returns no data, say so and suggest the user may not have access
   to that slice of data due to their role.
"""


def _build_system_prompt(
    schema_json: str,
    rls_username: str,
    user_filter_values: list[str],
) -> str:
    """Build the system prompt dynamically based on RLS config."""
    rls = get_rls_config()

    if rls.get("enabled") and user_filter_values:
        ft = rls["filter_table"]
        fc = rls["filter_column"]
        desc = rls.get("description", "")
        values_str = ", ".join(user_filter_values)
        rls_context = (
            f"## Current User RLS Context\n"
            f"The current user ({rls_username}) has access to these "
            f"{ft}.{fc} values: {values_str}\n"
            f"All DAX query results will be automatically filtered to these values.\n"
            f"{desc}"
        )
        rls_filter_rule = (
            f"Do NOT add {ft}/{fc} filters yourself — the system "
            f"automatically applies RLS filters via CALCULATETABLE + TREATAS."
        )
    elif rls.get("enabled"):
        rls_context = (
            f"## Current User RLS Context\n"
            f"The current user ({rls_username}) has no accessible data in this model."
        )
        rls_filter_rule = "The user has no data access. Inform them politely."
    else:
        rls_context = "## Note\nNo Row-Level Security is configured for this dataset."
        rls_filter_rule = "No RLS filters are applied. Queries return all data."

    return SYSTEM_PROMPT_BASE.format(
        schema_json=schema_json,
        rls_context=rls_context,
        rls_filter_rule=rls_filter_rule,
    )


def _extract_dax(text: str) -> str | None:
    """Extract a DAX query from a ```dax code fence."""
    m = re.search(r"```dax\s*\n(.*?)```", text, re.DOTALL | re.IGNORECASE)
    if m:
        return m.group(1).strip()
    # Fallback: look for EVALUATE anywhere
    m = re.search(r"(EVALUATE\b.+)", text, re.DOTALL | re.IGNORECASE)
    if m:
        return m.group(1).strip()
    return None


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


async def chat(
    user_message: str,
    rls_username: str,
    conversation_history: list[dict[str, str]] | None = None,
) -> dict[str, Any]:
    """
    Process a chat message:
      1. Send conversation + schema to LLM
      2. If LLM generates DAX → execute with RLS
      3. Feed results back to LLM for summarisation
      4. Return final answer + optional DAX + raw data
    """
    client = _get_openai_client()

    # Build schema context
    schema = await get_dataset_schema()

    # Get user's allowed filter values for the system prompt
    rls = get_rls_config()
    user_filter_values: list[str] = []
    if rls.get("enabled"):
        token = get_dax_token()
        user_filter_values = await get_user_filter_values(rls_username, token)

    system_msg = _build_system_prompt(
        schema_json=json.dumps(schema, indent=2),
        rls_username=rls_username,
        user_filter_values=user_filter_values,
    )

    messages: list[dict[str, str]] = [{"role": "system", "content": system_msg}]
    if conversation_history:
        messages.extend(conversation_history[-10:])
    messages.append({"role": "user", "content": user_message})

    # --- Step 1: Ask LLM to generate DAX (or conversational reply) -----------
    resp1 = await client.chat.completions.create(
        model=settings.azure_openai_deployment,
        messages=messages,
        temperature=0.1,
        max_tokens=1024,
    )
    assistant_text = resp1.choices[0].message.content or ""

    dax_query = _extract_dax(assistant_text)

    if dax_query is None:
        return {
            "answer": assistant_text,
            "dax": None,
            "data": None,
        }

    # --- Step 2: Execute DAX with RLS impersonation --------------------------
    try:
        rows = await execute_dax(dax_query, rls_username)
    except Exception as exc:
        logger.exception("DAX execution failed")
        return {
            "answer": f"I generated a query but it failed to execute: {exc}",
            "dax": dax_query,
            "data": None,
        }

    # --- Step 3: Summarise results -------------------------------------------
    result_text = json.dumps(rows[:200], indent=2)
    messages.append({"role": "assistant", "content": assistant_text})
    messages.append({
        "role": "user",
        "content": (
            f"Here are the query results (JSON). Summarise them in natural language "
            f"for the user. If tabular, show a Markdown table.\n\n{result_text}"
        ),
    })

    resp2 = await client.chat.completions.create(
        model=settings.azure_openai_deployment,
        messages=messages,
        temperature=0.3,
        max_tokens=1024,
    )
    final_answer = resp2.choices[0].message.content or ""

    return {
        "answer": final_answer,
        "dax": dax_query,
        "data": rows[:50],
    }

File: chat_engine.rb
require 'json'
require 'rest-client'
require_relative 'config'

module ChatEngine
  extend self

  def logger
    Settings.logger
  end

  def generate_dax(user_message, rls_username, conversation_history = [])
    schema = load_schema
    rls = Settings.rls_config
    fabric_rules = Settings.fabric_agent_instructions

    prompt = <<~PROMPT
      You are a Power BI DAX assistant.

      Goal:
      - If the user asks for dataset-backed analytics, return DAX in mode=dax.
      - If the user asks conversational or unsupported questions, return mode=answer.

      Rules for mode=dax:
      - Output a valid DAX query beginning with EVALUATE.
      - Prefer SUMMARIZECOLUMNS for grouped answers.
      - Always constrain by the current user using CALCULATETABLE with identity filter.
      - Use this identity filter when enabled:
        '#{rls['identity_table']}'[#{rls['identity_column']}] = "#{escape_dax_string(rls_username.to_s)}"
      - Use only tables/columns/measures from the schema.
      - When asked about hours, utilization, or expense types: ALWAYS group by BOTH
        'Dim_User'[Username] AND 'Dim_Task'[Expense Type] in SUMMARIZECOLUMNS.
        Include [Actual Hours] and [_Util+ %] as named measures in the result.
        Example pattern for hours/utilization by expense type:
          EVALUATE
          CALCULATETABLE(
              SUMMARIZECOLUMNS(
                  'Dim_User'[Username],
                  'Dim_Task'[Expense Type],
                  "Actual Hours", [Actual Hours],
                  "Util %", [_Util+ %]
              ),
              'Dim_User'[Username] = "<rls_username>"
          )
      - Never return only a single total — always break down by Expense Type.
      - Always include the date range that covers the data (use 'Date'[Date] bounds if filtering).

      Rules for mode=answer:
      - If the user sends a simple greeting (for example: hi, hello, hey), respond conversationally with exactly: "Hello! How can I help you today?"
      - For other non-data questions, keep the response short and helpful.

      Fabric-style business instructions to enforce:
      #{fabric_rules}

      Dataset schema (JSON):
      #{schema.to_json}

      RLS config (JSON):
      #{rls.to_json}

      Return ONLY strict JSON in one of these shapes:
      {"mode":"dax","dax":"EVALUATE ..."}
      {"mode":"answer","answer":"..."}
    PROMPT

    messages = build_messages(prompt, user_message, conversation_history)
    raw = azure_chat_completion(messages)
    parsed = parse_llm_json(raw)

    mode = parsed['mode'].to_s
    if mode == 'dax' && !parsed['dax'].to_s.strip.empty?
      { 'mode' => 'dax', 'dax' => parsed['dax'].to_s.strip }
    else
      answer = parsed['answer'].to_s.strip
      answer = 'I can help with data questions from your Power BI model. Please rephrase your request.' if answer.empty?
      { 'mode' => 'answer', 'answer' => answer }
    end
  rescue => e
    logger.error "generate_dax failed: #{e.class} - #{e.message}"
    raise e
  end

  def summarize(user_message, dax, results, conversation_history = [])
    rows = results.is_a?(Array) ? results : []
    preview = rows.first(Settings.summary_row_limit)
    fabric_rules = Settings.fabric_agent_instructions

    prompt = <<~PROMPT
      You are a data analyst assistant presenting Power BI query results.

      Formatting rules:
      - Always state the user's name/email and the date range covered by the data at the top.
      - When results contain 'Expense Type' or similar category breakdowns, present them as a
        Markdown table with User as the first column and each unique Expense Type as a column header.
        Fill each cell with the corresponding Actual Hours value (blank if none).
      - After the table, add a bullet list showing Utilization % (to 2 decimal places) for each
        Expense Type that has a non-null, non-zero Util % value.
      - If results are empty, explain that no data was found for the requested scope.
      - Do not invent values not present in the data.
      - Close with a friendly offer to provide a different date range or more detail.

      Fabric-style business instructions to also enforce:
      #{fabric_rules}
    PROMPT

    user_payload = {
      question: user_message,
      dax: dax,
      rows_total: rows.length,
      rows_preview: preview
    }

    messages = []
    messages << { role: 'system', content: prompt }
    conversation_history.last(8).each do |item|
      role = map_role(item['role'])
      content = item['content'].to_s
      next if content.empty?

      messages << { role: role, content: content }
    end
    messages << { role: 'user', content: "Summarize this result set: #{user_payload.to_json}" }

    azure_chat_completion(messages)
  rescue => e
    logger.error "summarize failed: #{e.class} - #{e.message}"
    raise e
  end

  private

  def load_schema
    path = File.join(File.dirname(__FILE__), 'sample_report', 'schema.json')
    if File.exist?(path)
      JSON.parse(File.read(path))
    else
      { 'tables' => [] }
    end
  rescue => e
    logger.warn "Failed loading schema.json: #{e.class} - #{e.message}"
    { 'tables' => [] }
  end

  def build_messages(system_prompt, user_message, history)
    messages = []
    messages << { role: 'system', content: system_prompt }
    history.last(8).each do |item|
      role = map_role(item['role'])
      content = item['content'].to_s
      next if content.empty?

      messages << { role: role, content: content }
    end
    messages << { role: 'user', content: user_message.to_s }
    messages
  end

  def map_role(role)
    case role.to_s
    when 'assistant', 'bot'
      'assistant'
    when 'system'
      'system'
    else
      'user'
    end
  end

  def azure_chat_completion(messages)
    endpoint = Settings.azure_openai_endpoint.to_s.chomp('/')
    deployment = Settings.azure_openai_deployment
    api_version = Settings.azure_openai_api_version

    raise 'AZURE_OPENAI_ENDPOINT is not configured' if endpoint.empty?
    raise 'AZURE_OPENAI_DEPLOYMENT is not configured' if deployment.to_s.strip.empty?

    url = "#{endpoint}/openai/deployments/#{deployment}/chat/completions?api-version=#{api_version}"

    headers = {
      content_type: :json,
      accept: :json
    }

    api_key = Settings.azure_openai_api_key.to_s
    if api_key.empty?
      token = Settings.azure_openai_bearer_token
      raise 'AZURE_OPENAI_API_KEY is missing and no Azure CLI token is available' if token.to_s.strip.empty?

      headers[:Authorization] = "Bearer #{token}"
    else
      headers['api-key'] = api_key
    end

    body = {
      messages: messages,
      temperature: 0.2,
      max_tokens: Settings.azure_openai_max_tokens
    }

    response = RestClient.post(url, body.to_json, headers)
    parsed = JSON.parse(response.body)
    content = parsed.dig('choices', 0, 'message', 'content').to_s
    raise 'Empty response from Azure OpenAI' if content.strip.empty?

    content
  rescue RestClient::ExceptionWithResponse => e
    logger.error "Azure OpenAI error #{e.response.code}: #{e.response.body[0..600]}"
    raise "Azure OpenAI request failed (#{e.response.code})"
  end

  def parse_llm_json(text)
    begin
      return JSON.parse(text)
    rescue JSON::ParserError
      # Best-effort recovery when the model wraps JSON in prose.
      start_idx = text.index('{')
      end_idx = text.rindex('}')
      raise "Invalid JSON response from model: #{text[0..200]}" if start_idx.nil? || end_idx.nil? || end_idx <= start_idx

      JSON.parse(text[start_idx..end_idx])
    end
  end

  def escape_dax_string(value)
    value.gsub('"', '""')
  end
end
